From 3cb3c50a27ab0748745bd0355bbcc796b473c8cf Mon Sep 17 00:00:00 2001
From: Bob Paauwe <bob.j.paauwe@intel.com>
Date: Mon, 7 Nov 2016 10:52:45 -0800
Subject: [PATCH 02/43] drm/i915/gen9: Add notifier callback for GPU frequency
 change

Whenever the GPU frequency is changed, notify any other kernel processes
that have registered to receive this information.

Signed-off-by: Bob Paauwe <bob.j.paauwe@intel.com>
Signed-off-by: Wan Ahmad Zainie <wan.ahmad.zainie.wan.mohamad@intel.com>
---
 drivers/gpu/drm/i915/intel_pm.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index d67c39f..92e9861 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -31,6 +31,7 @@
 #include "intel_drv.h"
 #include "../../../platform/x86/intel_ips.h"
 #include <linux/module.h>
+#include <linux/notifier.h>
 #include <drm/drm_atomic_helper.h>
 
 #if IS_ENABLED(CONFIG_DRM_I915_GVT)
@@ -40,6 +41,8 @@
 #define GEN9_TURBO_DISABLED    0
 #define GEN9_TURBO_ENABLED     1
 
+static BLOCKING_NOTIFIER_HEAD(i915_freq_notifier_list);
+
 /**
  * DOC: RC6
  *
@@ -6222,6 +6225,9 @@ static int gen6_set_rps(struct drm_i915_private *dev_priv, u8 val)
 				   GEN6_FREQUENCY(val) |
 				   GEN6_OFFSET(0) |
 				   GEN6_AGGRESSIVE_TURBO);
+
+		blocking_notifier_call_chain(&i915_freq_notifier_list,
+					     (unsigned long)val, NULL);
 	}
 
 	/* Make sure we continue to get interrupts
@@ -7850,6 +7856,26 @@ void intel_gpu_tfm_teardown(void)
 }
 
 /**
+ * Register a notifier callback for gpu frequency changes.
+ * @nb: pointer to the notifier block for the callback
+ */
+int i915_register_freq_notify(struct notifier_block *nb)
+{
+	return blocking_notifier_chain_register(&i915_freq_notifier_list, nb);
+}
+EXPORT_SYMBOL_GPL(i915_register_freq_notify);
+
+/**
+ * Unregister a notifier from the gpu freqency change callback.
+ * @nb: pointer to the notifier block for the callback
+ */
+int i915_unregister_freq_notify(struct notifier_block *nb)
+{
+	return blocking_notifier_chain_unregister(&i915_freq_notifier_list, nb);
+}
+EXPORT_SYMBOL_GPL(i915_unregister_freq_notify);
+
+/**
  * i915_gpu_disable_turbo - disable grahics turbo
  *
  * Restrict the GPU frequencies to only those below the tfm limit.
-- 
1.9.1

